<h1>Interpolación

In [2]:
import numpy as np
import sympy as sym
import csv
import wget
import os.path as path

Raíces de polinomios

1.

El error es de tipo $O(x_i^2)$ debido al truncamiento realizado al usar la expansión de Taylor. Como x_i es escojida por aleatoriamente, la magnitud del error puede ser muy masiva si la adivinanza es muy leja al resultado matemático.

2.

Se puede ajustar la presición cambiando el valor de $\epsilon$. A menor valor $\epsilon$ mayor presición.

3.

In [3]:
e=10**-6
x=sym.Symbol('x')
def raphson(x1):
    def f(x):
        return(3*x**5+5*x**4-x**3)
    def df(x):
        return(15*x**4+20*x**3-3*x**2)
    xn=x1-(f(x1)/df(x1))
    if e >= abs(xn-x1)/abs(xn):
        return(xn)
    else:
        return(raphson(xn))

In [4]:
print(raphson(100))

0.18046042171637047


La ecuación $3*x^5+5*x^4-x^3$ tiene una raiz en $\approx$ 0.18046042171637047. Por lo menos desde este valor hasta 100 no hay más raices.

4.

In [13]:
def rootlegendre(ni,inter):
    e=10**-6
    x=sym.Symbol('x')
    n=sym.Symbol('n')
    def L(ni):
        return(sym.diff((x**2-1)**n,x,ni)*1/(2**ni*sym.factorial(ni)))
    xm=(inter[0]+inter[1])/2
    lmax=L(ni).subs(n,ni).subs(x,inter[0]).evalf()
    lmin=L(ni).subs(n,ni).subs(x,inter[1]).evalf()
    lmid=L(ni).subs(n,ni).subs(x,xm).evalf()
    if lmax*lmin < 0:
        if lmid == 0:
            return(xm)
        elif L(ni).subs(n,ni).subs(x,inter[0]).evalf()*L(ni).subs(n,ni).subs(x,xm).evalf() < 0:
            return(rootlegendre(ni,[inter[0],xm]))
        else:
            return(rootlegendre(ni,[xm,inter[1]]))
    else:
        return("No hay raices en el intervalo según este método")

inter=[-1+e,1-e]
for i in range(21):
    a=rootlegendre(i,inter)
    print(["Valor de n",i,"Valor de la raiz",a])

['Valor de n', 0, 'Valor de la raiz', 'No hay raices en el intervalo según este método']
['Valor de n', 1, 'Valor de la raiz', 0.0]
['Valor de n', 2, 'Valor de la raiz', 'No hay raices en el intervalo según este método']
['Valor de n', 3, 'Valor de la raiz', 0.0]
['Valor de n', 4, 'Valor de la raiz', 'No hay raices en el intervalo según este método']
['Valor de n', 5, 'Valor de la raiz', 0.0]
['Valor de n', 6, 'Valor de la raiz', 'No hay raices en el intervalo según este método']
['Valor de n', 7, 'Valor de la raiz', 0.0]
['Valor de n', 8, 'Valor de la raiz', 'No hay raices en el intervalo según este método']
['Valor de n', 9, 'Valor de la raiz', 0.0]
['Valor de n', 10, 'Valor de la raiz', 'No hay raices en el intervalo según este método']
['Valor de n', 11, 'Valor de la raiz', 0.0]
['Valor de n', 12, 'Valor de la raiz', 'No hay raices en el intervalo según este método']
['Valor de n', 13, 'Valor de la raiz', 0.0]
['Valor de n', 14, 'Valor de la raiz', 'No hay raices en el intervalo se

Curiosamente, aunque existen más raices para los polinomios, el método es incapaz de encontrarlas debido a la condición de $f(a)*f(b)<0$

\*además, se utilizó un intervalo (-1,1) en vez de \[1,1\] por que la ecuación se intermina en -1 y 1

pero si usamos sympy.solvers.solve(), nos da los valores exactos, pero con algunos quedando expresados en formas de "Rootof". Sin embargo, esos pueden ser evaluados uno por uno. Esto, lamentablemente, aumenta el timepo de procesamiento.

In [ ]:
def rootlegendre(ni,inter):
    e=10**-6
    x=sym.Symbol('x')
    n=sym.Symbol('n')
    def L(ni):
        return(sym.diff((x**2-1)**n,x,ni)*1/(2**ni*sym.factorial(ni)))
    a=[]
    for i in sym.solvers.solve(L(ni).subs(n,ni),x):
        val=i.evalf()
        if(inter[1]>val and inter[0]<val):
            a.append(val)
    return(a)
inter=[-1,1]
for i in range(21):
    a=rootlegendre(i,inter)
    print(["Valor de n",i,"tiene raiz",a])
    print("")
    print("")

5.

In [ ]:
def rootlaguerre(ni,inter):
    e=10**-6
    x=sym.Symbol('x')
    n=sym.Symbol('n')
    def L(ni):
        return(sym.diff((sym.E**(-x)*x**n)**n,x,ni)*sym.E**(x)/(sym.factorial(ni)))
        a=[]
    for i in sym.solvers.solve(L(ni).subs(n,ni),x):
        a.append(i.evalf())
    return(a)
inter=[0,100000]
for i in range(20):
    a=rootlaguerre(i,inter)
    print(["Valor de n",i,"Valor de la raiz",a])

['Valor de n', 0, 'Valor de la raiz', [-0.993128599185095, -0.963971927277914, -0.912234428251326, -0.839116971822219, -0.746331906460151, -0.636053680726515, -0.510867001950827, -0.373706088715420, -0.227785851141645, -0.0765265211334973, 0.0765265211334973, 0.227785851141645, 0.373706088715420, 0.510867001950827, 0.636053680726515, 0.746331906460151, 0.839116971822219, 0.912234428251326, 0.963971927277914, 0.993128599185095]]
['Valor de n', 1, 'Valor de la raiz', [-0.993128599185095, -0.963971927277914, -0.912234428251326, -0.839116971822219, -0.746331906460151, -0.636053680726515, -0.510867001950827, -0.373706088715420, -0.227785851141645, -0.0765265211334973, 0.0765265211334973, 0.227785851141645, 0.373706088715420, 0.510867001950827, 0.636053680726515, 0.746331906460151, 0.839116971822219, 0.912234428251326, 0.963971927277914, 0.993128599185095, 1.00000000000000]]
['Valor de n', 2, 'Valor de la raiz', [-0.993128599185095, -0.963971927277914, -0.912234428251326, -0.839116971822219,

Interpolación

1.

La ecuación del polinomio interpolador de Lagrange es:

\begin{equation}
    p_n(x)=\sum^n_{i=0}f(x_i)L_i(x)
\end{equation}

dónde...

\begin{equation}
    L_i(x)=\prod^n_{j=0,j \neq i}\frac{x-x_j}{x_i-x_j}
\end{equation}

Para que dos ecuaciones tengan el mismo polinomio de lagrange necesitarian el mismo $L_i(x)$, pero $L_i(x)$ de pende de $x_i$, $x_j$ y $x$. $x$ es una variable, así que su valor es arbitrario. $x_i$ y $x_j$ representan el set de datos en el eje X. Así que solo otro set de datos identico al usado puede tener el mismo polinomio.

Esto significa que el polinomio es ÚNICO PARA UN SET DE DATOS DADO

4.

In [1]:
archivo='Parabolico.csv'
url='https://raw.githubusercontent.com/asegura4488/Database/main/MetodosComputacionalesReforma/Parabolico.csv'
if not (path.exists(archivo)):
    Path_ = wget.download(url,archivo)
else:
    print('--Archivo ya existe---')
    Path_ = archivo

--Archivo ya existe---


In [7]:
with open('Parabolico.csv', newline='') as f:
    reader = csv.reader(f)
    data = list(reader)
X=[]
Y=[]
i=1
while i < len(data):
    X.append(float(data[i][0]))
    Y.append(float(data[i][1]))
    i=i+1

[1.4, 3.5, 5.6]

In [10]:
def lagrange(x,xi,j):
    prod = 1.0
    n = len(xi)
    
    for i in range(n):
        if i != j:
            prod *= (x - xi[i])/(xi[j]-xi[i])
            
    return prod
def poly(x,xi,yi):
    
    Sum = 0.
    n = len(xi)
        
    for j in range(n):
        Sum += yi[j]*lagrange(x,xi,j)
        
    return Sum

In [61]:
x=sym.Symbol('x')
y=poly(x,X,Y)
f=poly(x,X,Y)
f=sym.expand(f)
f

-0.0554912422401579*x**2 + 0.363970234266202*x

Usando las propiedades de un tiro parabolico se obtiene que:

$9.40t\approx x$

osea:

$V_{0y}=0.363970234266202*t\approx 3.42$ \\

y

$V_{0x}=9.40$

In [62]:
V0y=3.42
V0x=9.40
mag=np.sqrt(V0x**2+V0x**2)
ang=np.arctan(V0y/V0x)*360/(2*np.pi)
print(mag,ang)

13.293607486307094 19.992893979869876


5.

In [66]:
archivo='InterpolacionNewtonNoequi.csv'
url='https://raw.githubusercontent.com/asegura4488/Database/main/MetodosComputacionalesReforma/InterpolacionNewtonNoequi.csv'
if not (path.exists(archivo)):
    Path_ = wget.download(url,archivo)
else:
    print('--Archivo ya existe---')
    Path_ = archivo

--Archivo ya existe---


In [67]:
with open('InterpolacionNewtonNoequi.csv', newline='') as f:
    reader = csv.reader(f)
    data = list(reader)
X=[]
Y=[]
i=1
while i < len(data):
    X.append(float(data[i][0]))
    Y.append(float(data[i][1]))
    i=i+1

In [138]:
def fn(X,Y,i,n):
    if n==0:
        return(Y[i])
    else:
        a=fn(X,Y,i,n-1)
        b=fn(X,Y,i-1,n-1)
        res=(a-b)/(X[i]-X[0])
        return(res)

def pn(X,Y,i):
    if i==0:
        return(fn(X,Y,i,i))
    else:
        var=sym.Symbol('x')
        f=1
        for j in range(i):
           f=f*(var-X[j]) 
        return(sym.expand(pn(X,Y,i-1)+fn(X,Y,i,i)*f))            

In [169]:
M=[]
for i in range(6):
    M.append(pn(X,Y,i))
for i in M:
    print(i)

-18.0
3.33333333333333*x - 18.0
1.58333333333333*x**2 + 0.958333333333333*x - 18.0
-0.775684016134519*x**3 + 4.29822738980415*x**2 - 1.36871871507022*x - 18.0
0.237838826695891*x**4 - 2.51190745101452*x**3 + 8.17500026494717*x**2 - 4.08008133940338*x - 18.0
-0.043275194660857*x**5 + 0.735503565295746*x**4 - 4.54411059228837*x**3 + 11.6309573105632*x**2 - 6.15209765976521*x - 18.0


Hasta el grado 5 (maximo grado permitido por la ecuación) todavía todos los coedicientes son mayores a 0. Por ende el menor grado de la ecuación es grado 5.